shared/domains/chelonia/GIMessage.js

```diff
@@ -38,59 +53,97 @@ export class GIMessage {
   static OP_PROTOCOL_UPGRADE: 'pu' = 'pu'
   static OP_PROP_SET: 'ps' = 'ps' // set a public key/value pair
   static OP_PROP_DEL: 'pd' = 'pd' // delete a public key/value pair
+  static OP_CONTRACT_AUTH: 'ca' = 'ca' // authorize a contract
+  static OP_CONTRACT_DEAUTH: 'cd' = 'cd' // deauthorize a contract
+  static OP_ATOMIC: 'at' = 'at' // atomic op
+  static OP_KEYSHARE: 'ks' = 'ks' // key share
+  static OP_KEY_REQUEST: 'kr' = 'kr' // key request
+  static OP_KEY_REQUEST_RESPONSE: 'krr' = 'krr' // key request response
 
   // eslint-disable-next-line camelcase
   static createV1_0 (
-    contractID: string | null = null,
-    previousHEAD: string | null = null,
-    op: GIOp,
-    manifest: string,
-    signatureFn?: Function = defaultSignatureFn
+    {
+      contractID,
+      originatingContractID,
+      previousHEAD = null,
+      op,
+      manifest,
+      signatureFn = defaultSignatureFn
+    }: {
+      contractID: string | null,
+      originatingContractID?: string,
+      previousHEAD?: string | null,
+      op: GIOp,
+      manifest: string,
+      signatureFn?: Function
+    }
   ): this {
-    const message = {
+    const head = {
       version: '1.0.0',
       previousHEAD,
       contractID,
-      op,
-      manifest,
-      // the nonce makes it difficult to predict message contents
-      // and makes it easier to prevent conflicts during development
-      nonce: Math.random()
+      originatingContractID,
+      op: op[0],
+      manifest
     }
+    console.log('createV1_0', { op, head })
+    const message = op[1]
     // NOTE: the JSON strings generated here must be preserved forever.
     //       do not ever regenerate this message using the contructor.
     //       instead store it using serialize() and restore it using
     //       deserialize().
+    const headJSON = JSON.stringify(head)
     const messageJSON = JSON.stringify(message)
+    const signedPayload = blake32Hash(`${blake32Hash(headJSON)}${blake32Hash(messageJSON)}`)
+    const signature = signatureFn(signedPayload)
     const value = JSON.stringify({
+      head: headJSON,
       message: messageJSON,
-      sig: signatureFn(messageJSON)
+      sig: signature
     })
     return new this({
       mapping: { key: blake32Hash(value), value },
-      message
+      head,
+      message,
+      decryptedValue: op.length === 3 ? op[2] : op[1],
```

Comment:
Why is `decryptedValue` always being set?

Reply:
This is the relevant commit: https://github.com/okTurtles/group-income/commit/2f70afa7314195136b4b789987b9546ce43962a7